### PR TITLE
feat(Slider): add option to add Tooltip over thumb

### LINKED ISSUE #5955: Slider: can we provide a tooltip for the thumb like PF3 did?
Screen Shot 2021-04-23 at 11 52 21

### PR DESCRIPTION
**Pull request by KKoukiou**
Fixes #5955

### CODE REVIEW

File: packages/react-core/src/components/Slider/examples/Slider.md [hunk 1]
```diff
@@ -134,13 +135,22 @@ class ContinuousInput extends React.Component {
     return (
       <>
         <Text component={TextVariants.h3}>Slider value is: {this.state.value.toFixed(2)}</Text>
```
Comment (nicolethoen) [inline packages/react-core/src/components/Slider/examples/Slider.md:137]: We should remove this line.

File: packages/react-core/src/components/Slider/examples/Slider.md [hunk 2]
```diff
@@ -134,13 +135,22 @@ class ContinuousInput extends React.Component {
...
+          label="thumbHasTooltip"
+          isChecked={this.state.thumbHasTooltip}
+          onChange={thumbHasTooltip => this.setState({ thumbHasTooltip })} />
+        <Text component={TextVariants.h3}>Slider Value is: {this.state.value.toFixed(2)}</Text>
+        <Slider
+          thumbHasTooltip={this.state.thumbHasTooltip}
+          value={this.state.value}
+          onChange={this.onChange} />
         <br />
         <Text component={TextVariants.h3}>Slider value is: {this.state.valueCustom.toFixed(2)}</Text>
```
Comment (nicolethoen) [inline packages/react-core/src/components/Slider/examples/Slider.md:148]: As it is, the `Slide value is:` text in the example only has 2 values trailing the decimal, but the tooltip has many. I wonder if we should make the number of trailing values something the user can configure, and have it always default to 2? It'd make the aria label much more readable.

**Review by nicolethoen — CHANGES_REQUESTED** [review]
Could you also add a unit test which basically verifies that a Tooltip gets rendered whenever `thumbHasTooltip` is true?

File: packages/react-core/src/components/Slider/examples/Slider.md (hunk 2 above)
Comment (KKoukiou) [reply packages/react-core/src/components/Slider/examples/Slider.md:147]: Made it default to 2 decimals. Lets add the property for configuring once someone requests it if you 're fine with that.

**Comment by KKoukiou** [comment]
> Could you also add a unit test which basically verifies that a Tooltip gets rendered whenever `thumbHasTooltip` is true?

Added

File: packages/react-core/src/components/Slider/examples/Slider.md (hunk 1 above)
Comment (KKoukiou) [reply packages/react-core/src/components/Slider/examples/Slider.md]: Removed.

File: packages/react-core/src/components/Slider/examples/Slider.md (hunk 2 above)
Comment (nicolethoen) [reply packages/react-core/src/components/Slider/examples/Slider.md:147]: Could you also remove the trimming in the examples if it's no longer needed?

Comment (KKoukiou) [reply packages/react-core/src/components/Slider/examples/Slider.md]: Done

**Review by mcarrano — CHANGES_REQUESTED** [review]
This looks good, although while the tooltip in the example is constrained to 2 decimal places, the value displayed is not. Shouldn't these match? Here's what I'm seeing.

Screen Shot 2021-11-22 at 10 34 21 AM

Also, it seems like the Value input examples are broken. Maybe unrelated to your changes, but can you check these?

File: packages/react-core/src/components/Slider/Slider.tsx [hunk 3]
```diff
@@ -139,7 +143,9 @@ export const Slider: React.FunctionComponent<SliderProps> = ({
         return matchingStep.label;
       }
     }
-    return localValue.toString();
+    // For continuous steps default to showing 2 decimals in tooltip
+    // Consider making it configurable via a property
+    return Number(localValue.toFixed(2)).toString();
```
Comment (nicolethoen) [inline packages/react-core/src/components/Slider/Slider.tsx:148]: To piggy pack off of Matt's comment, I think we want to make this same update to the `onChangeHandler` function. Either that, or in the `handleThumbMove` function we should trim the continuous value on line 221 to use 2 trailing digits.

Comment (KKoukiou) [reply packages/react-core/src/components/Slider/Slider.tsx:148]: Fixed.

**Comment by KKoukiou** [comment]
> This looks good, although while the tooltip in the example is constrained to 2 decimal places, the value displayed is not. Shouldn't these match? Here's what I'm seeing.
> 
> Screen Shot 2021-11-22 at 10 34 21 AM
> 
> Also, it seems like the Value input examples are broken. Maybe unrelated to your changes, but can you check these?

This is also broken on master, just checked. Let's fix it in a separate PR as it's unrelated.

File: packages/react-core/src/components/Slider/examples/Slider.md [hunk 4]
```diff
@@ -133,14 +134,23 @@ class ContinuousInput extends React.Component {
   render() {
     return (
       <>
-        <Text component={TextVariants.h3}>Slider value is: {this.state.value.toFixed(2)}</Text>
-        <Slider value={this.state.value} onChange={this.onChange} />
+        <Checkbox
+          id="thumb-has-tooltip"
```
Comment (gabipodolnikova) [inline packages/react-core/src/components/Slider/examples/Slider.md:138]: I would add a style={{ marginBottom: 20 }} to the Checkbox.

Comment (KKoukiou) [reply packages/react-core/src/components/Slider/examples/Slider.md:138]: Added.

**Review by gabipodolnikova — CHANGES_REQUESTED** [review]
I also noticed that the Thumb value input example is broken, which it wasn't before. When I move it, I get: `TypeError: W.toFixed is not a function` Do you see it too?

**Comment by KKoukiou** [comment]
> I also noticed that the Thumb value input example is broken, which it wasn't before. When I move it, I get: `TypeError: W.toFixed is not a function` Do you see it too?

Yes, sorry. I fixed it.

**Review by gabipodolnikova — APPROVED** [review]

**Review by nicolethoen — APPROVED** [review]

**Comment by KKoukiou** [comment]
> Tooltip option looks good, but some of the Discrete slider examples seem to be broken - specifically the first one and the first two Vale input examples. I cannot drag the slider.

This problem exists already upstream, see on the docs pge directly without my PR https://www.patternfly.org/v4/components/slider/

**Review by mcarrano — APPROVED** [review]
Fair enough @KKoukiou . In that case I will approve this PR, but @nicolethoen @tlabaj do you have any idea when or why these examples broke. We should try to clean them up for the next release. Do we need a new issue for this?

**Comment by KKoukiou** [comment]
@mcarrano I created an issue last week about it: https://github.com/patternfly/patternfly-react/issues/6644

File: packages/react-core/src/components/Slider/Slider.tsx [hunk 5]
```diff
@@ -44,6 +45,8 @@ export interface SliderProps extends Omit<React.HTMLProps<HTMLDivElement>, 'onCh
   inputAriaLabel?: string;
   /* Aria label for the thumb */
   thumbAriaLabel?: string;
+  /* Adds a tooltip over the thumb containing the current value */
+  thumbHasTooltip?: boolean;
```
Comment (tlabaj) [inline packages/react-core/src/components/Slider/Slider.tsx:49]: Can we just call this `hasTooltip`.  This is more consistent with our naming practices.  There is also no other tooltip so the name will not conflict.

**Review by tlabaj — CHANGES_REQUESTED** [review]
Looks great! One suggestion.

File: packages/react-core/src/components/Slider/Slider.tsx (hunk 5 above)
Comment (nicolethoen) [reply packages/react-core/src/components/Slider/Slider.tsx:49]: @KKoukiou @tlabaj 
I think referencing _what_ has the tooltip in the prop name is wise - maybe `hasTooltipOverThumb` is descriptive and follows the naming patterns?

Comment (KKoukiou) [reply packages/react-core/src/components/Slider/Slider.tsx]: Fixed! Sorry for the wrong time to reply here.

**Review by tlabaj — APPROVED** [review]
Looks Good!  Thank you!
Looks like you need to update the snapshots and we can merge.

**Pull request merged**